exchanges/subscription/list.go

```diff
@@ -30,3 +59,186 @@ func (l List) GroupPairs() (n List) {
 }
+
+// QualifiedChannels returns a sorted list of all the Qualified Channels in the list
+func (l List) QualifiedChannels() []string {
+	c := make([]string, len(l))
+	for i := range l {
+		c[i] = l[i].QualifiedChannel
+	}
+	slices.Sort(c)
+	return c
+}
+
+// SetStates sets the state for all the subs in a list
+// Errors are collected for any subscriptions already in the state
+// On error all changes are reverted
+func (l List) SetStates(state State) error {
+	var err error
+	for _, sub := range l {
+		err = common.AppendError(err, sub.SetState(state))
+	}
+	return err
+}
+
+func fillAssetPairs(ap assetPairs, a asset.Item, e iExchange) error {
+	p, err := e.GetEnabledPairs(a)
+	if err != nil {
+		return err
+	}
+	f, err := e.GetPairFormat(a, true)
+	if err != nil {
+		return err
+	}
+	ap[a] = p.Format(f)
+	return nil
+}
+
+// AssetPairs returns a map of enabled pairs for the subscriptions in the list, formatted for the asset
+func (l List) AssetPairs(e iExchange) (assetPairs, error) { //nolint:revive // unexported-return is acceptable since it's a undecorated primitive
+	at := e.GetAssetTypes(true)
+	ap := assetPairs{}
+	for _, s := range l {
+		switch s.Asset {
+		case asset.Empty:
+			// Nothing to do
+		case asset.All:
+			for _, a := range at {
+				if err := fillAssetPairs(ap, a, e); err != nil {
+					return nil, err
+				}
+			}
+		default:
+			if slices.Contains(at, s.Asset) {
+				if err := fillAssetPairs(ap, s.Asset, e); err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+	return ap, nil
+}
+
+type tplCtx struct {
+	Sub        *Subscription
+	AssetPairs assetPairs
+	Assets     asset.Items
+}
+
+/*
+ExpandTemplates returns a list of Subscriptions with Template expanded
+Format of the Template should be text/template compatible.
+Template Variables:
+  - $s is the subscription; e.g. {{$s.Interval}} {{$s.Params.freq}}
+  - $asset will fan out the enabled assets
+    sub.Asset must be All; Otherwise just hardcode the asset the subscription is for
+  - $pair will fan out the pairs for the assets, formatted for request
+    Must be used in cojunction with $asset when Asset is All, otherwise we don't know what pairs to use
+    May not be used when Asset is Empty
+
+Calls e.GetSubscriptionTemplateFuncs for a template.FuncMap for flexibility in pipelines, e.g. {{ assetName "$asset" }}
+Filters out Authenticated subscriptions if CanUseAuthenticatedEndpoints is false
+*/
+func (l List) ExpandTemplates(e iExchange) (List, error) {
+	if !e.CanUseAuthenticatedWebsocketEndpoints() {
+		n := List{}
+		for _, s := range l {
+			if !s.Authenticated {
+				n = append(n, s)
+			}
+		}
+		l = n
+	}
+
+	ap, err := l.AssetPairs(e)
+	if err != nil {
+		return nil, err
+	}
+
+	assets := make([]asset.Item, 0, len(ap))
+	for k := range ap {
+		assets = append(assets, k)
+	}
+
+	baseTpl := template.New("channel")
+	if funcs := e.GetSubscriptionTemplateFuncs(); funcs != nil {
+		baseTpl = baseTpl.Funcs(funcs)
+	}
+
+	subs := List{}
+
+	rs := "\x1E"
```

Comment:
could be a const here or global const

Reply:
Yep. I actually went hunting for it as a const in go, and was surprised to not find it.

Fixed 51e7c49